rllib/policy/policy_map.py

```diff
@@ -66,130 +52,97 @@ def __init__(
-            observation_space: The observation space of the
-                policy.
-            action_space: The action space of the policy.
-            merged_config: The config object (or complete config dict) for the policy
-                to use.
-        """
-        if config_override is not None:
-            deprecation_warning(
-                old="PolicyMap.create_policy(config_override=..)",
-                error=True,
-            )
-        _class = get_tf_eager_cls_if_necessary(policy_cls, merged_config)
-
-        policy = create_policy_for_framework(
-            policy_id,
-            _class,
-            merged_config,
-            observation_space,
-            action_space,
-            self.worker_index,
-            self.session_creator,
-            self.seed,
-        )
-        self.insert_policy(policy_id, policy)
+        # Ray object store references to the stashed Policy states.
+        self._policy_state_refs = {}
 
     @with_lock
     @override(dict)
     def __getitem__(self, item):
         # Never seen this key -> Error.
         if item not in self.valid_keys:
-            raise KeyError(f"PolicyID '{item}' not found in this PolicyMap!")
+            raise KeyError(
+                f"PolicyID '{item}' not found in this PolicyMap! "
+                f"IDs stored in this map: {self.valid_keys}."
+            )
 
-        # Item already in cache -> Rearrange deque (least recently used) and
-        # return.
+        # Item already in cache -> Rearrange deque (promote `item` to
+        # "most recently used") and return it.
         if item in self.cache:
             self.deque.remove(item)
             self.deque.append(item)
-        # Item not currently in cache -> Get from disk and - if at capacity -
+            return self.cache[item]
+
+        # Item not currently in cache -> Get from stash and - if at capacity -
         # remove leftmost one.
         else:
-            self._read_from_disk(policy_id=item)
-
-        return self.cache[item]
+            assert item in self._policy_state_refs
+            policy_state = ray.get(self._policy_state_refs[item])
+
+            # All our policies have same NN-architecture (are "swappable").
+            # -> Get least recently used one, stash its state to disk and load
+            # new policy's state into that one. This way, we save the costly
+            # re-creation step.
+            if self.policies_swappable and len(self.deque) == self.deque.maxlen:
+                old_policy_id = self.deque[0]
+                policy = self.cache[old_policy_id]
+                self._stash_least_used_policy()
+                self.cache[item] = policy
+                # Restore policy's state and return it.
+                policy.set_state(policy_state)
+            # Policies are different or we are not at capacity:
+            # Have to (re-)create new policy here.
+            else:
+                self._stash_least_used_policy()
```

Comment:
Looks like we are always stashing if self.policies_swappable and **not** len(self.deque) == self.deque.maxlen.
Meaning we always stash as long as our deque is not full.

Reply:
In case of a full deque, we already call _stash_least_used_policy in __setitem__

Reply:
I think you are right! Great catch. Let me fix this ...